packages/shylu/basker/src/basker_solve_rhs.hpp

```diff
@@ -838,140 +553,138 @@ namespace BaskerNS
-	    y(gperm(jnew)) -= M.val(i)*x(k+bcol);
-	    
-	  }
+        const Int j = M.row_idx(i) + msrow;
+
+        //y(gperm(j)) -= M.val(i)*x(k+bcol); //NDE: Why is gperm needed here? x is const
+        y(j) -= M.val(i)*xkbcol; //NDE: Why is gperm needed here? x is const
       }
-    
+    }
+
     return 0;
   }//neg_spmv
 
 
-
   //M\x = y
   template <class Int, class Entry, class Exe_Space>
   BASKER_INLINE
   int Basker<Int,Entry,Exe_Space>::lower_tri_solve
   (
    BASKER_MATRIX &M,
-   ENTRY_1DARRAY x, 
-   ENTRY_1DARRAY y
+   ENTRY_1DARRAY &x, //NDE rhs after MOD
+   ENTRY_1DARRAY &y  //NDE 0's/soln after MOD
    )
   {
     const Int bcol = M.scol;
     const Int brow = M.scol;
-    
+
     //M.info();
 
     //printf("Lower-Tri-Solve-Test, [%d %d %d %d] \n",
     //	  M.srow, M.nrow, M.scol, M.ncol);
 
     for(Int k = 0; k < M.ncol; ++k)
+    {
+      //Test if zero pivot value
+#ifdef BASKER_DEBUG_SOLVE_RHS
+      BASKER_ASSERT(M.val[M.col_ptr[k]]!=0.0, "LOWER PIVOT 0");
+#endif
+      /*
+      if(M.val[M.col_ptr[k]] == 0.0) 
       {
-	//Test if zero pivot value
-	#ifdef BASKER_DEBUG_SOLVE_RHS
-	BASKER_ASSERT(M.val[M.col_ptr[k]]!=0.0, "LOWER PIVOT 0");
-	#endif
-	
-	if(M.val[M.col_ptr[k]] == 0.0)
-	  {
-	    printf("Lower Pivot: %d %f \n", 
-		   M.row_idx[M.col_ptr[k]],
-		   M.val[M.col_ptr[k]]);
-	    return -1;
-	  }
-
-	//printf("Lower tri.  k: %d out: %f in: %f piv: %f \n",
-	//   k+bcol, y[k+bcol], x[k+bcol], M.val[M.col_ptr[k]]);
-
-	//Replace with Entry divide in future
-	//y[k+bcol] = x[k+bcol] / M.val[M.col_ptr[k]];
-
-	y(k+brow) = x(k+bcol) / M.val(M.col_ptr(k));
-	
-	//for(Int i = M.col_ptr[k]+1; i < M.col_ptr[k+1]; i++)
-	for(Int i = M.col_ptr(k)+1; i < M.col_ptr(k+1); ++i)
-	  {
-	   
-
-	    //Int j = gperm[M.row_idx[i]];
-	    const Int j = gperm(M.row_idx(i)+brow);
-	    
-	    #ifdef BASKER_DEBUG_SOLVE_RHS
-	    BASKER_ASSERT(j != BASKER_MAX_IDX,"Using nonperm\n");
-	    #endif
-	    
-	    //x[j] -= M.val[i]*y[k+bcol];
-	    //printf("gperm: %d x(%d) y(i)  \n",
-	    //	   M.row_idx(i) + brow, j, k+bcol);
-	    x(j) -= M.val(i)*y(k+bcol);
-
-	  }//over all nnz in a column
-
-      }//over each column
+        printf("Lower Pivot: %d %f \n", 
+            M.row_idx[M.col_ptr[k]],
+            M.val[M.col_ptr[k]]);
+        return -1;
+      }
+      */
 
-    return 0;
+      //printf("Lower tri.  k: %d out: %f in: %f piv: %f \n",
+      //   k+bcol, y[k+bcol], x[k+bcol], M.val[M.col_ptr[k]]);
+
+      //Replace with Entry divide in future
+      //y[k+bcol] = x[k+bcol] / M.val[M.col_ptr[k]];
+
+      const Int istart = M.col_ptr(k);
+      const Int iend = M.col_ptr(k+1);
+
+      y(k+brow) = x(k+bcol) / M.val(M.col_ptr(k));
+
+      const auto ykbcol = y(k+bcol);
+      //for(Int i = M.col_ptr(k)+1; i < M.col_ptr(k+1); ++i)
+      for(Int i = istart+1; i < iend; ++i)
+      {
+        //Int j = gperm[M.row_idx[i]];
+//        const Int j = gperm(M.row_idx(i)+brow); //NDE: why use gperm?
```

Comment:
basker->Options.no_pivot      = BASKER_TRUE; can be used to check if gperm is needed or not